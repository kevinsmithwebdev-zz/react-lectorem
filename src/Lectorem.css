.Lectorem {
  align-items: center;
  display: flex;
  flex-direction: column;
  margin: 5px;
  padding: 5px;
  background-color: #fff;
  width: 800px;
}

audio {
  filter: invert(20%);
  width: 400px;
  height: 50px;
}

.Lectorem .playerContainer {
  padding: 5px;
}

.Lectorem-paragraph-header {
  text-align: center;
  margin: 5px;
  padding: 5px;
}

.Lectorem .paragraph {
  text-align: left;
  margin: 5px 10px;;
  padding: 5px;
}

.Lectorem h2 {
  margin: 5px;
  padding: 0;
}

.Lectorem h3 {
  margin: 5px;
  padding: 0;
}
