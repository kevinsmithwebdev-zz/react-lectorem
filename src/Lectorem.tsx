import React, { useState } from 'react';
import Player from './Player/Player';
import classes from './Lectorem.module.css';
import Modal from './Modal/Modal';
import Texts from './Texts/Texts';
import { StoryDataInterface, ConfigurationInterface, LectInterface } from './interfaces/index';

interface ComponentPropsInterface {
  data: StoryDataInterface,
  configuration?: ConfigurationInterface,
}

export const createShowModal =
  (setModalData: Function): Function =>
    (lect: LectInterface): void => (lect.translation || lect.explanation) && setModalData(lect);

export const createHideModal =
  (setModalData: Function): Function =>
    (): void => setModalData(null);

const Lectorem: React.FC<ComponentPropsInterface> = ({ data, configuration }) => {
  const [modalData, setModalData] = useState(null);
  const [readTime, setReadTime] = useState(null);

  if (!data) {
    return <p>Loading...</p>;
  }

  const showModal = createShowModal(setModalData);
  const hideModal = createHideModal(setModalData);

  const {
    title, subtitle, audioPath, paragraphs,
  } = data;

  return (
    <div className={ classes.Lectorem }>
      { !!title && <h2>{title} </h2> }
      { !!subtitle && <h3>{subtitle} </h3> }
      <div className={ classes.playerContainer }>
        <Player audioPath={ audioPath } onListen={ setReadTime } />
      </div>
      <div className={ classes.lectsContainer }>
        <Texts
          configuration={ configuration }
          paragraphs={ paragraphs }
          readTime={ readTime }
          showModal={ showModal }
        />
      </div>

      <Modal data={ modalData } hideModal={ hideModal } />
    </div>
  );
}

export default Lectorem;
