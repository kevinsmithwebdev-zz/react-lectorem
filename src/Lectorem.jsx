import React, { Component } from 'react';
import Player from './Player/Player';
import './Lectorem.css';
import Modal from './Modal/Modal';
import Paragraphs from './Paragraphs/Paragraphs';

// TODO: issues with packages using hooks
class Lectorem extends Component {
  state = {
    modalData: null,
    readTime: 0,
  }

  showModal = lect => (lect.translation || lect.explanation) && this.setState({ modalData: lect });

  hideModal = () => this.setState({ modalData: null });

  setReadTime = readTime => this.setState({ readTime });

  render() {
    const { data } = this.props;

    if (!data) {
      return <p>Loading...</p>;
    }

    const {
      title, subtitle, audioPath, paragraphs,
    } = data;

    return (
      <div className='Lectorem'>
        { !!title && <h2>{ title } </h2> }
        { !!subtitle && <h3>{ subtitle } </h3> }
        <div className='playerContainer'>
          <Player
            audioPath={ audioPath }
            onListen={ this.setReadTime }
          />
        </div>
        <div className='lectsContainer'>
          <Paragraphs
            paragraphs={ paragraphs }
            readTime={ this.state.readTime }
            showModal={ this.showModal }
          />
        </div>

        <Modal
          data={ this.state.modalData }
          hideModal={ this.hideModal }
        />

      </div>
    );
  }
}

export default Lectorem;
