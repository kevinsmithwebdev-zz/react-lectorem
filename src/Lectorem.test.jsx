import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import cloneDeep from 'lodash/cloneDeep';
import Lectorem, { createShowModal, createHideModal } from './Lectorem';
import storyData from '../fixture/storyData.json';

describe('Lectorem', () => {
  describe('the main component render', () => {
    const renderer = new ShallowRenderer();
    describe('empty', () => {
      it('should render without crashing', () => {
        renderer.render(<Lectorem />);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
      });
    });

    describe('with undefined data', () => {
      const props = { data: undefined };
      it('should render without crashing', () => {
        renderer.render(<Lectorem {...props} />);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
      });
    });

    describe('with data', () => {
      describe('with all data', () => {
        const props = { data: storyData };
        it('should render without crashing', () => {
          renderer.render(<Lectorem {...props} />);
          expect(renderer.getRenderOutput()).toMatchSnapshot();
        });
      });

      describe('with data missing title', () => {
        const data = cloneDeep(storyData);
        delete data.title;
        const props = { data };
        it('should render without crashing', () => {
          renderer.render(<Lectorem {...props} />);
          expect(renderer.getRenderOutput()).toMatchSnapshot();
        });
      });

      describe('with data missing subtitle', () => {
        const data = cloneDeep(storyData);
        delete data.subtitle;
        const props = { data };
        it('should render without crashing', () => {
          renderer.render(<Lectorem {...props} />);
          expect(renderer.getRenderOutput()).toMatchSnapshot();
        });
      });
    });
  });

  describe('createShowModal', () => {
    const setModalData = jest.fn();
    const showModal = createShowModal(setModalData);
    describe('the function it returns', () => {
      describe('when lect has translation but no explanation', () => {
        const lect = { translation: 'translation123' };
        beforeAll(() => {
          setModalData.mockReset();
          showModal(lect);
        });
        it('should call setModalData with correct lect', () => {
          expect(setModalData).toBeCalledWith(lect);
        });
      });

      describe('when lect no translation but has explanation', () => {
        const lect = { explanation: 'explanation123' };
        beforeAll(() => {
          setModalData.mockReset();
          showModal(lect);
        });
        it('should call setModalData with correct lect', () => {
          expect(setModalData).toBeCalledWith(lect);
        });
      });

      describe('when lect has neither translation nor explanation', () => {
        const lect = { other: 'data' };
        beforeAll(() => {
          setModalData.mockReset();
          showModal(lect);
        });
        it('should call not call setModalData ', () => {
          expect(setModalData).not.toBeCalled();
        });
      });
    });
  });

  describe('createHideModal', () => {
    const setModalData = jest.fn();
    const hideModal = createHideModal(setModalData);
    describe('the function it returns', () => {
      beforeAll(() => {
        setModalData.mockReset();
        hideModal();
      });
      it('should call setModalData with null', () => {
        expect(setModalData).toBeCalledWith(null);
      });
    });
  });
});
