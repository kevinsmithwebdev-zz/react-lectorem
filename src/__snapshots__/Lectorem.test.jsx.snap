// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lectorem the main component render empty should render without crashing 1`] = `
<p>
  Loading...
</p>
`;

exports[`Lectorem the main component render with data with all data should render without crashing 1`] = `
<div
  className="Lectorem"
>
  <h2>
    El burro bailarín
     
  </h2>
  <h3>
    The Dancing Donkey
     
  </h3>
  <div
    className="playerContainer"
  >
    <Player
      audioPath="./burro.mp3"
      onListen={[Function]}
    />
  </div>
  <div
    className="lectsContainer"
  >
    <Paragraphs
      paragraphs={
        Array [
          Object {
            "header": Object {
              "imageFilename": "burro.png",
              "imagePath": "./burro.png",
              "lects": Array [
                Object {
                  "end": 1.18,
                  "start": 0.36,
                  "text": "El burro",
                },
                Object {
                  "end": 2.58,
                  "explanation": "The Spanish word \\"bailarín\\" can be both and adjective (\\"dancing\\") and a noun (\\"dancer\\").",
                  "start": 1.18,
                  "text": "bailarín.",
                  "translation": "dancer",
                },
              ],
            },
            "lects": Array [
              Object {
                "end": 3.22,
                "explanation": "This is how we say hello.",
                "start": 2.64,
                "text": "Hola.",
              },
              Object {
                "end": 4.12,
                "explanation": "",
                "start": 3.6,
                "text": "Soy",
                "translation": "I am",
              },
              Object {
                "end": 4.98,
                "explanation": "",
                "start": 4.12,
                "text": "un burro.",
                "translation": "a donkey.",
              },
              Object {
                "end": 6.36,
                "explanation": "",
                "start": 5.42,
                "text": "Me llamo",
                "translation": "My name is",
              },
              Object {
                "end": 7.04,
                "explanation": "",
                "start": 6.36,
                "text": "Beto.",
                "translation": "Beto.",
              },
              Object {
                "end": 8.36,
                "explanation": "",
                "start": 7.7,
                "text": "Tengo",
                "translation": "I have",
              },
              Object {
                "end": 8.86,
                "explanation": "",
                "start": 8.4,
                "text": "botas.",
                "translation": "boots.",
              },
              Object {
                "end": 9.92,
                "explanation": "",
                "start": 9.46,
                "text": "Tengo",
                "translation": "I have",
              },
              Object {
                "end": 10.52,
                "explanation": "",
                "start": 9.92,
                "text": "botas",
                "translation": "boots",
              },
              Object {
                "end": 11.6,
                "explanation": "",
                "start": 10.68,
                "text": "para bailar.",
                "translation": "for dancing.",
              },
            ],
          },
        ]
      }
      readTime={0}
      showModal={[Function]}
    />
  </div>
  <Modal
    data={null}
    hideModal={[Function]}
  />
</div>
`;

exports[`Lectorem the main component render with data with data missing subtitle should render without crashing 1`] = `
<div
  className="Lectorem"
>
  <h2>
    El burro bailarín
     
  </h2>
  <div
    className="playerContainer"
  >
    <Player
      audioPath="./burro.mp3"
      onListen={[Function]}
    />
  </div>
  <div
    className="lectsContainer"
  >
    <Paragraphs
      paragraphs={
        Array [
          Object {
            "header": Object {
              "imageFilename": "burro.png",
              "imagePath": "./burro.png",
              "lects": Array [
                Object {
                  "end": 1.18,
                  "start": 0.36,
                  "text": "El burro",
                },
                Object {
                  "end": 2.58,
                  "explanation": "The Spanish word \\"bailarín\\" can be both and adjective (\\"dancing\\") and a noun (\\"dancer\\").",
                  "start": 1.18,
                  "text": "bailarín.",
                  "translation": "dancer",
                },
              ],
            },
            "lects": Array [
              Object {
                "end": 3.22,
                "explanation": "This is how we say hello.",
                "start": 2.64,
                "text": "Hola.",
              },
              Object {
                "end": 4.12,
                "explanation": "",
                "start": 3.6,
                "text": "Soy",
                "translation": "I am",
              },
              Object {
                "end": 4.98,
                "explanation": "",
                "start": 4.12,
                "text": "un burro.",
                "translation": "a donkey.",
              },
              Object {
                "end": 6.36,
                "explanation": "",
                "start": 5.42,
                "text": "Me llamo",
                "translation": "My name is",
              },
              Object {
                "end": 7.04,
                "explanation": "",
                "start": 6.36,
                "text": "Beto.",
                "translation": "Beto.",
              },
              Object {
                "end": 8.36,
                "explanation": "",
                "start": 7.7,
                "text": "Tengo",
                "translation": "I have",
              },
              Object {
                "end": 8.86,
                "explanation": "",
                "start": 8.4,
                "text": "botas.",
                "translation": "boots.",
              },
              Object {
                "end": 9.92,
                "explanation": "",
                "start": 9.46,
                "text": "Tengo",
                "translation": "I have",
              },
              Object {
                "end": 10.52,
                "explanation": "",
                "start": 9.92,
                "text": "botas",
                "translation": "boots",
              },
              Object {
                "end": 11.6,
                "explanation": "",
                "start": 10.68,
                "text": "para bailar.",
                "translation": "for dancing.",
              },
            ],
          },
        ]
      }
      readTime={0}
      showModal={[Function]}
    />
  </div>
  <Modal
    data={null}
    hideModal={[Function]}
  />
</div>
`;

exports[`Lectorem the main component render with data with data missing title should render without crashing 1`] = `
<div
  className="Lectorem"
>
  <h3>
    The Dancing Donkey
     
  </h3>
  <div
    className="playerContainer"
  >
    <Player
      audioPath="./burro.mp3"
      onListen={[Function]}
    />
  </div>
  <div
    className="lectsContainer"
  >
    <Paragraphs
      paragraphs={
        Array [
          Object {
            "header": Object {
              "imageFilename": "burro.png",
              "imagePath": "./burro.png",
              "lects": Array [
                Object {
                  "end": 1.18,
                  "start": 0.36,
                  "text": "El burro",
                },
                Object {
                  "end": 2.58,
                  "explanation": "The Spanish word \\"bailarín\\" can be both and adjective (\\"dancing\\") and a noun (\\"dancer\\").",
                  "start": 1.18,
                  "text": "bailarín.",
                  "translation": "dancer",
                },
              ],
            },
            "lects": Array [
              Object {
                "end": 3.22,
                "explanation": "This is how we say hello.",
                "start": 2.64,
                "text": "Hola.",
              },
              Object {
                "end": 4.12,
                "explanation": "",
                "start": 3.6,
                "text": "Soy",
                "translation": "I am",
              },
              Object {
                "end": 4.98,
                "explanation": "",
                "start": 4.12,
                "text": "un burro.",
                "translation": "a donkey.",
              },
              Object {
                "end": 6.36,
                "explanation": "",
                "start": 5.42,
                "text": "Me llamo",
                "translation": "My name is",
              },
              Object {
                "end": 7.04,
                "explanation": "",
                "start": 6.36,
                "text": "Beto.",
                "translation": "Beto.",
              },
              Object {
                "end": 8.36,
                "explanation": "",
                "start": 7.7,
                "text": "Tengo",
                "translation": "I have",
              },
              Object {
                "end": 8.86,
                "explanation": "",
                "start": 8.4,
                "text": "botas.",
                "translation": "boots.",
              },
              Object {
                "end": 9.92,
                "explanation": "",
                "start": 9.46,
                "text": "Tengo",
                "translation": "I have",
              },
              Object {
                "end": 10.52,
                "explanation": "",
                "start": 9.92,
                "text": "botas",
                "translation": "boots",
              },
              Object {
                "end": 11.6,
                "explanation": "",
                "start": 10.68,
                "text": "para bailar.",
                "translation": "for dancing.",
              },
            ],
          },
        ]
      }
      readTime={0}
      showModal={[Function]}
    />
  </div>
  <Modal
    data={null}
    hideModal={[Function]}
  />
</div>
`;

exports[`Lectorem the main component render with undefined data should render without crashing 1`] = `
<p>
  Loading...
</p>
`;
